blockchain/message_pool/src/msgpool.rs

```diff
@@ -0,0 +1,1066 @@
+        let msg_serial = msg.marshal_cbor()?;
+        self.add(msg).await?;
+        self.add_local(msg, msg_serial)?;
+        msg.cid().map_err(|err| err.into())
+    }
+
+    /// This is a helper to push that will help to make sure that the message fits the parameters
+    /// to be pushed to the MessagePool
+    pub async fn add(&mut self, msg: &SignedMessage) -> Result<(), Error> {
+        let size = msg.marshal_cbor()?.len();
+        if size > 32 * 1024 {
+            return Err(Error::MessageTooBig);
+        }
+        if msg.value() > &BigUint::from(2_000_000_000u64) {
+            return Err(Error::MessageValueTooHigh);
+        }
+
+        self.verify_msg_sig(msg)?;
+        let tmp = msg.clone();
+
+        let tip_locked = self.cur_tipset.read().await;
+        let tip = tip_locked.clone();
+        drop(tip_locked);
+
+        self.add_tipset(tmp, &tip).await
+    }
+
+    /// Add a SignedMessage without doing any of the checks
+    pub async fn add_skip_checks(&mut self, m: SignedMessage) -> Result<(), Error> {
+        self.add_helper(m).await
+    }
+
+    /// Verify the message signature. first check if it has already been verified and put into
+    /// cache. If it has not, then manually verify it then put it into cache for future use
+    fn verify_msg_sig(&mut self, msg: &SignedMessage) -> Result<(), Error> {
+        let cid = msg.cid()?;
+
+        if let Some(()) = self.sig_val_cache.get(&cid) {
+            return Ok(());
+        }
+
+        let umsg = msg.message().marshal_cbor()?;
+        msg.signature()
+            .verify(umsg.as_slice(), msg.from())
+            .map_err(Error::Other)?;
+
+        self.sig_val_cache.put(cid, ());
+
+        Ok(())
+    }
+
+    /// Verify the state_sequence and balance for the sender of the message given then call add_locked
+    /// to finish adding the signed_message to pending
+    async fn add_tipset(&mut self, msg: SignedMessage, cur_ts: &Tipset) -> Result<(), Error> {
+        let sequence = self.get_state_sequence(msg.from(), cur_ts).await?;
+
+        if sequence > msg.message().sequence() {
+            return Err(Error::SequenceTooLow);
+        }
+
+        let balance = self.get_state_balance(msg.from(), cur_ts).await?;
+
+        let msg_balance = BigInt::from(msg.message().required_funds());
+        if balance < msg_balance {
+            return Err(Error::NotEnoughFunds);
+        }
+        self.add_helper(msg).await
+    }
+
+    /// Finish verifying signed message before adding it to the pending mset hashmap. If an entry
+    /// in the hashmap does not yet exist, create a new mset that will correspond to the from message
+    /// and push it to the pending phashmap
+    async fn add_helper(&mut self, msg: SignedMessage) -> Result<(), Error> {
+        add_helper(
+            self.api.clone(),
+            self.bls_sig_cache.clone(),
+            self.pending.clone(),
+            msg,
+        )
+        .await
```

Comment:
in head_change there is a call to this function, and since head_change is used in an async block that doesn't have access to mpool, just those specific parts of it, it made sense to just have self.head_change act as a wrapper method.

Reply:
okay so why does head_change not use self? seems like it probably should instead of passing in all the fields you need?

Reply:
whoops, slight mistake in my last comment that I corrected. head_change can't use self because the async block that calls it doesn't have access to self, if it did, then all of the mpool would have to be wrapped in an RwLock, which I wanted to avoid.